src/bench/connectblock.cpp

```diff
@@ -0,0 +1,145 @@
+ * - All Taproot inputs use simple key path spends (no script path spends)
+ * - All signatures use SIGHASH_ALL (default sighash)
+ * - Each transaction spends all outputs from the previous transaction
+ */
+CBlock CreateTestBlock(
+    TestChain100Setup& test_setup,
+    const std::vector<CKey>& keys,
+    const std::vector<CTxOut>& outputs,
+    int num_txs = 1000)
+{
+    Chainstate& chainstate{test_setup.m_node.chainman->ActiveChainstate()};
+
+    const WitnessV1Taproot coinbase_taproot{XOnlyPubKey(test_setup.coinbaseKey.GetPubKey())};
+
+    // Create the outputs that will be spent in the first transaction of the test block
+    // Doing this in a separate blocks excludes the validation of it's inputs from the benchmark
+    auto coinbase_to_spend{test_setup.m_coinbase_txns[0]};
+    const auto [first_tx, _]{test_setup.CreateValidTransaction(
+        std::vector{coinbase_to_spend},
+        std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
+        chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+    const auto test_block_parent_coinbase{GetScriptForDestination(coinbase_taproot)};
+    test_setup.CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
+
+    std::vector<CMutableTransaction> txs;
+    txs.reserve(num_txs);
+    CTransactionRef input_tx{MakeTransactionRef(first_tx)};
+    for (int i{0}; i < num_txs; i++) {
+        std::vector<COutPoint> inputs;
+        inputs.reserve(outputs.size());
+
+        for (size_t j{0}; j < outputs.size(); j++) {
+            inputs.emplace_back(input_tx->GetHash(), j);
+        }
+
+        const auto [taproot_tx, _]{test_setup.CreateValidTransaction(
+            std::vector{input_tx}, inputs, chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+        txs.emplace_back(taproot_tx);
+        input_tx = MakeTransactionRef(taproot_tx);
+    }
+
+    // Coinbase output can use any output type as it is not spent and will not change the benchmark
+    const CScript coinbase_spk{GetScriptForDestination(coinbase_taproot)};
+    return test_setup.CreateBlock(txs, coinbase_spk, chainstate);
+}
+
+/*
+ * Creates key pairs and corresponding outputs for the benchmark transactions.
+ * - For Taproot outputs: Creates simple key path spendable outputs
+ * - For non-Taproot outputs: Creates P2WPKH (native SegWit v0) outputs
+ * - All outputs have value of 1 BTC
+ */
+std::pair<std::vector<CKey>, std::vector<CTxOut>> CreateKeysAndOutputs(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
+{
+    std::vector<CKey> keys{coinbaseKey};
+    keys.reserve(num_taproot + num_nontaproot + 1);
+
+    std::vector<CTxOut> outputs;
+    outputs.reserve(num_taproot + num_nontaproot);
+
+    for (size_t i{0}; i < num_nontaproot; i++) {
+        const CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{key.GetPubKey()}));
```

Comment:
to avoid possible copying we could take advantage of `emplace_back` constructing the key in-place:
```suggestion
        keys.emplace_back(GenerateRandomKey());
        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{keys.back().GetPubKey()}));
```